<h1 style="color:#0c0a0b;background-color:#71b12c;font-size: 5rem; text-align: center;padding:0.5rem;border-radius:5rem;  border-bottom: 1.5rem solid #edf2f8"> Quiz 1: Fibonacci</h1>

In [1]:
%%time
import math

class FibonacciAnalyzer:
    def __init__(self, a0=1, a1=1):
        """
        Initialize Fibonacci sequence with custom starting values
        Default: a0 = a1 = 1 (standard Fibonacci)
        """
        self.a0 = a0
        self.a1 = a1
        self.phi = (1 + math.sqrt(5)) / 2  # Golden ratio exact value
        
    def fibonacci_term(self, n):
        """Calculate nth Fibonacci term"""
        if n == 0:
            return self.a0
        elif n == 1:
            return self.a1
        
        prev_prev = self.a0
        prev = self.a1
        
        for i in range(2, n + 1):
            current = prev + prev_prev
            prev_prev = prev
            prev = current
            
        return prev
    
    def ratio_approximation(self, n):
        """Calculate a_{n+1}/a_n ratio"""
        if n == 0:
            return self.a1 / self.a0
        
        a_n = self.fibonacci_term(n)
        a_n_plus_1 = self.fibonacci_term(n + 1)
        
        return a_n_plus_1 / a_n
    
    def relative_error(self, n):
        """
        Calculate relative error as defined:
        |a_{n+1}/a_n - a_n/a_{n-1}|
        """
        if n < 1:
            return None
            
        ratio_current = self.ratio_approximation(n)
        ratio_previous = self.ratio_approximation(n - 1)
        
        return abs(ratio_current - ratio_previous)
    
    def find_n_for_error_threshold(self, error_threshold=1e-6):
        """Find the value of n where relative error <= threshold"""
        n = 1
        max_iterations = 100
        
        print(f"Searching for n where relative error <= {error_threshold}")
        print("n\tRatio\t\tRel Error\tPhi Error")
        print("-" * 60)
        
        while n < max_iterations:
            rel_error = self.relative_error(n)
            ratio = self.ratio_approximation(n)
            phi_error = abs(ratio - self.phi)
            
            print(f"{n}\t{ratio:.10f}\t{rel_error:.2e}\t{phi_error:.2e}")
            
            if rel_error and rel_error <= error_threshold:
                return n
                
            n += 1
            
        return None
    
    def analyze_arbitrary_start(self, a0, a1, max_terms=20):
        """Analyze what happens with arbitrary starting values"""
        print(f"\nAnalyzing Fibonacci sequence with a0={a0}, a1={a1}")
        print("n\tFib(n)\t\tRatio\t\tLimit→φ Error")
        print("-" * 55)
        
        # Temporarily change starting values
        original_a0, original_a1 = self.a0, self.a1
        self.a0, self.a1 = a0, a1
        
        ratios = []
        for n in range(max_terms):
            fib_n = self.fibonacci_term(n)
            if n > 0:
                ratio = self.ratio_approximation(n-1)
                phi_error = abs(ratio - self.phi)
                ratios.append(ratio)
                print(f"{n}\t{fib_n}\t\t{ratio:.8f}\t{phi_error:.2e}")
            else:
                print(f"{n}\t{fib_n}")
        
        # Restore original values
        self.a0, self.a1 = original_a0, original_a1
        
        # Check if ratios converge to phi
        if len(ratios) > 5:
            last_ratio = ratios[-1]
            convergence_error = abs(last_ratio - self.phi)
            print(f"\nFinal ratio: {last_ratio:.10f}")
            print(f"φ = {self.phi:.10f}")
            print(f"Convergence error: {convergence_error:.2e}")
            
            if convergence_error < 1e-6:
                print("✓ Sequence converges to φ")
            else:
                print("⚠ Sequence may not converge to φ (needs more terms)")
        
        return ratios
    
    def mathematical_analysis(self):
        """Answer the mathematical questions"""
        print("=" * 60)
        print("MATHEMATICAL ANALYSIS")
        print("=" * 60)
        
        print(f"1. Exact value of φ (Golden Ratio):")
        print(f"   φ = (1 + √5)/2 = {self.phi:.15f}")
        
        print(f"\n2. φ satisfies the equation: φ² = φ + 1")
        phi_squared = self.phi ** 2
        phi_plus_one = self.phi + 1
        print(f"   φ² = {phi_squared:.15f}")
        print(f"   φ + 1 = {phi_plus_one:.15f}")
        print(f"   Difference: {abs(phi_squared - phi_plus_one):.2e}")
        
        print(f"\n3. For ANY starting values a₀, a₁ ∈ ℕ:")
        print("   The limit lim(n→∞) aₙ₊₁/aₙ = φ")
        print("   This is because the ratio of consecutive terms")
        print("   in ANY Fibonacci-like sequence converges to φ")
        print("   (as long as the ratio a₁/a₀ is positive)")

def main():
    # Create analyzer with standard Fibonacci sequence
    analyzer = FibonacciAnalyzer()
    
    print("FIBONACCI GOLDEN RATIO ANALYSIS")
    print("=" * 60)
    
    # Question 1: Find n for relative error <= 10^-6
    print("1. Finding n for relative error ≤ 10⁻⁶:")
    n_threshold = analyzer.find_n_for_error_threshold(1e-6)
    
    if n_threshold:
        print(f"\n✓ Answer: n = {n_threshold}")
        print(f"At n = {n_threshold}, relative error ≤ 10⁻⁶")
    else:
        print("Could not find n within reasonable iterations")
    
    # Mathematical analysis
    analyzer.mathematical_analysis()
    
    print("\n" + "=" * 60)
    print("4. TESTING WITH ARBITRARY STARTING VALUES")
    print("=" * 60)
    
    # Test with different starting values
    test_cases = [
        (2, 3),
        (5, 8),
        (1, 4),
        (10, 15)
    ]
    
    for a0, a1 in test_cases:
        analyzer.analyze_arbitrary_start(a0, a1, 15)
        print()

if __name__ == "__main__":
    main()

FIBONACCI GOLDEN RATIO ANALYSIS
1. Finding n for relative error ≤ 10⁻⁶:
Searching for n where relative error <= 1e-06
n	Ratio		Rel Error	Phi Error
------------------------------------------------------------
1	2.0000000000	1.00e+00	3.82e-01
2	1.5000000000	5.00e-01	1.18e-01
3	1.6666666667	1.67e-01	4.86e-02
4	1.6000000000	6.67e-02	1.80e-02
5	1.6250000000	2.50e-02	6.97e-03
6	1.6153846154	9.62e-03	2.65e-03
7	1.6190476190	3.66e-03	1.01e-03
8	1.6176470588	1.40e-03	3.87e-04
9	1.6181818182	5.35e-04	1.48e-04
10	1.6179775281	2.04e-04	5.65e-05
11	1.6180555556	7.80e-05	2.16e-05
12	1.6180257511	2.98e-05	8.24e-06
13	1.6180371353	1.14e-05	3.15e-06
14	1.6180327869	4.35e-06	1.20e-06
15	1.6180344478	1.66e-06	4.59e-07
16	1.6180338134	6.34e-07	1.75e-07

✓ Answer: n = 16
At n = 16, relative error ≤ 10⁻⁶
MATHEMATICAL ANALYSIS
1. Exact value of φ (Golden Ratio):
   φ = (1 + √5)/2 = 1.618033988749895

2. φ satisfies the equation: φ² = φ + 1
   φ² = 2.618033988749895
   φ + 1 = 2.618033988749895
   Difference:


### 1. **Para qué valor de $ n $ obtenemos la estimación con error relativo $ \leq 10^{-6} $?**

Del análisis numérico realizado con el script, se observa que el error relativo cumple la condición $ \leq 10^{-6} $ para $ n = 16 $. 

**Resultado numérico:**
$
\begin{array}{|c|c|c|c|}
\hline
n & \text{Ratio } (a_{n+1}/a_n) & \text{Error Relativo} & \text{Error respecto a } \phi \\
\hline
16 & 1.6180338134 & 6.34 \times 10^{-7} & 1.75 \times 10^{-7} \\
\hline
\end{array}
$

**Razón matemática:**
El error relativo entre ratios consecutivos $ \left| \frac{a_{n+2}}{a_{n+1}} - \frac{a_{n+1}}{a_n} \right| $ decae exponencialmente hacia 0, ya que la secuencia de ratios converge a $ \phi $ con orden de convergencia lineal. Esto se debe a que el sistema dinámico subyacente tiene un valor propio dominante $ \phi $ y otro menor $ \psi = 1 - \phi $.



### 2. **¿Cuál es el valor exacto de $ \phi $?**

El valor exacto del número áureo $ \phi $ es la solución positiva de la ecuación cuadrática:
$
\phi^2 = \phi + 1
$
Despejando, obtenemos:
$
\phi = \frac{1 + \sqrt{5}}{2} \approx 1.618033988749895\ldots
$

**Verificación en el script:**
$
\phi_{\text{exacto}} = 1.618033988749895 \quad \text{(calculado con } \sqrt{5} \text{ en alta precisión)}
$

**Propiedad clave:**
$
\phi^2 = \phi + 1 \quad \Rightarrow \quad \phi = 1 + \frac{1}{\phi}
$

---

### 3. **¿Qué ocurre con el límite si tomamos $ a_0, a_1 \in \mathbb{N} $ arbitrarios?**

**Teorema:**
Para cualquier par de valores iniciales $ a_0, a_1 \in \mathbb{N} $ (no ambos cero), el límite de los ratios consecutivos satisface:
$
\lim_{n \to \infty} \frac{a_{n+1}}{a_n} = \phi
$

**Demostración intuitiva:**
- La recurrencia $ a_{n+1} = a_n + a_{n-1} $ tiene solución general:
  $
  a_n = A \phi^n + B \psi^n \quad \text{donde } \psi = 1 - \phi \approx -0.618.
  $
- Como $ |\psi| < 1 $, el término $ B \psi^n $ decae exponencialmente. Para $ n $ grande:
  $
  \frac{a_{n+1}}{a_n} \approx \frac{A \phi^{n+1}}{A \phi^n} = \phi.
  $

**Verificación numérica:**
El script muestra que incluso para $ (a_0, a_1) = (1, 4) $ o $ (10, 15) $, los ratios convergen a $ \phi $, aunque la velocidad de convergencia puede variar.

**Ejemplo con $ (a_0, a_1) = (1, 4) $:**
$
\begin{array}{|c|c|}
\hline
n & \text{Error } |\text{ratio} - \phi| \\
\hline
15 & 4.25 \times 10^{-6} \\
\hline
\end{array}
$

